com.hazelcast.client.impl.spi.impl.ClientInvocationTest

_master_ (commit 5f414fa3ad03417e0b9078c8d9f249904003c936)

Failed on openj9-17: https://jenkins.hazelcast.com/job/Hazelcast-master-Openj9-17/69/testReport/com.hazelcast.client.impl.spi.impl/ClientInvocationTest/invokeOnMemberRedirectsToRandom_whenMemberIsNotInMemberList/
<details><summary>Stacktrace:</summary>

```
java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "java.lang.invoke.MethodHandle.invokeExact(java.lang.invoke.MethodHandle, java.lang.Object[])"
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.returnOrThrowWithGetConventions(InvocationFuture.java:121)
	at com.hazelcast.client.impl.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:95)
	at com.hazelcast.client.impl.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:40)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:618)
	at com.hazelcast.client.impl.spi.impl.ClientInvocationTest.invokeOnMemberRedirectsToRandom_whenMemberIsNotInMemberList(ClientInvocationTest.java:213)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:858)
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.invoke.MethodHandle.invokeExact(java.lang.invoke.MethodHandle, java.lang.Object[])"
	at java.lang.StringConcatHelper.stringOf(StringConcatHelper.java)
	at com.hazelcast.map.impl.recordstore.DefaultRecordStore.getStateMessage(DefaultRecordStore.java:1411)
	at com.hazelcast.map.impl.recordstore.DefaultRecordStore.startLoading(DefaultRecordStore.java:1333)
	at com.hazelcast.map.impl.PartitionContainer.lambda$new$0(PartitionContainer.java:59)
	at com.hazelcast.internal.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:140)
	at com.hazelcast.map.impl.PartitionContainer.getRecordStore(PartitionContainer.java:150)
	at com.hazelcast.map.impl.operation.MapOperation.getRecordStoreOrNull(MapOperation.java:312)
	at com.hazelcast.map.impl.operation.MapOperation.beforeRun(MapOperation.java:108)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:269)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:219)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:180)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:144)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.loop(OperationThread.java:134)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.executeRun(OperationThread.java:115)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:111)
	at ------ submitted from ------.()
	at com.hazelcast.internal.util.ExceptionUtil.cloneExceptionWithFixedAsyncStackTrace(ExceptionUtil.java:348)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.returnOrThrowWithGetConventions(InvocationFuture.java:112)
	... 16 more

```
</details>


<details><summary>Standard output:</summary>

```
Finished Running Test: invokeOnPartitionOwnerRedirectsToRandom_WhenPartitionOwnerIsnull in 0.556 seconds.

No metrics recorded during the test

```
</details>

Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-master-Openj9-17/69/

The similar issue was for`invokeOnPartitionOwnerRedirectsToRandom_WhenPartitionOwnerIsnull` test

--- a/hazelcast/src/main/java/com/hazelcast/internal/util/FutureUtil.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/util/FutureUtil.java
@@ -24,6 +24,7 @@ import com.hazelcast.spi.annotation.PrivateApi;
 import com.hazelcast.spi.impl.InternalCompletableFuture;
 import com.hazelcast.transaction.TransactionTimedOutException;
 
+import javax.annotation.Nonnull;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -448,11 +449,12 @@ public final class FutureUtil {
     /**
      * Get all futures that are done
      *
-     * @param futures
+     * @param futures collection of futures which we will check for done futures.
      * @return list of completed futures
      */
+    @Nonnull
     public static List<Future<?>> getAllDone(Collection<Future<?>> futures) {
-        List<Future<?>> doneFutures = new ArrayList<Future<?>>();
+        List<Future<?>> doneFutures = new ArrayList<>();
         for (Future<?> f : futures) {
             if (f.isDone()) {
                 doneFutures.add(f);

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
@@ -88,6 +88,7 @@ import static com.hazelcast.map.impl.mapstore.MapDataStores.EMPTY_MAP_DATA_STORE
 import static com.hazelcast.map.impl.record.Record.UNSET;
 import static com.hazelcast.map.impl.recordstore.StaticParams.PUT_BACKUP_PARAMS;
 import static com.hazelcast.spi.impl.merge.MergingValueFactory.createMergingEntry;
+import static java.util.Collections.emptyList;
 
 /**
  * Default implementation of a record store.
@@ -1294,7 +1295,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         }
 
         if (FutureUtil.allDone(loadingFutures)) {
-            List<Future<?>> doneFutures = null;
+            List<Future<?>> doneFutures = emptyList();
             try {
                 doneFutures = FutureUtil.getAllDone(loadingFutures);
                 // check all finished loading futures for exceptions
@@ -1406,9 +1407,20 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
     }
 
     private String getStateMessage() {
-        return "on partitionId=" + partitionId + " on " + mapServiceContext.getNodeEngine().getThisAddress()
-                + " loadedOnCreate=" + loadedOnCreate + " loadedOnPreMigration=" + loadedOnPreMigration
-                + " isLoaded=" + isLoaded();
+        // due to weird issue with OpenJ9 implementation of string concatenation:
+        //noinspection StringBufferReplaceableByString
+        StringBuilder sb = new StringBuilder();
+        sb.append("on partitionId=");
+        sb.append(partitionId);
+        sb.append(" on ");
+        sb.append(mapServiceContext.getNodeEngine().getThisAddress());
+        sb.append(" loadedOnCreate=");
+        sb.append(loadedOnCreate);
+        sb.append(" loadedOnPreMigration=");
+        sb.append(loadedOnPreMigration);
+        sb.append(" isLoaded=");
+        sb.append(isLoaded());
+        return sb.toString();
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
@@ -93,7 +93,7 @@ import static java.util.Collections.emptyList;
 /**
  * Default implementation of a record store.
  */
-@SuppressWarnings({"checkstyle:methodcount", "checkstyle:classfanoutcomplexity"})
+@SuppressWarnings({"checkstyle:methodcount", "checkstyle:classfanoutcomplexity", "rawtypes"})
 public class DefaultRecordStore extends AbstractEvictableRecordStore {
     protected final ILogger logger;
     protected final RecordStoreLoader recordStoreLoader;
@@ -729,7 +729,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
 
         // then try to load missing keys from map-store
         if (mapDataStore != EMPTY_MAP_DATA_STORE && !keys.isEmpty()) {
-            Map loadedEntries = loadEntries(keys, callerAddress);
+            Map<Data, Object> loadedEntries = loadEntries(keys, callerAddress);
             addToMapEntrySet(mapEntries, loadedEntries);
         }
 
@@ -812,8 +812,8 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         mapEntries.add(dataKey, dataValue);
     }
 
-    public void addToMapEntrySet(MapEntries mapEntries, Map<Object, Object> entries) {
-        for (Map.Entry<Object, Object> entry : entries.entrySet()) {
+    public void addToMapEntrySet(MapEntries mapEntries, Map<?, ?> entries) {
+        for (Map.Entry<?, ?> entry : entries.entrySet()) {
             addToMapEntrySet(entry.getKey(), entry.getValue(), mapEntries);
         }
     }
@@ -1059,6 +1059,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public boolean merge(MapMergeTypes<Object, Object> mergingEntry,
                          SplitBrainMergePolicy<Object, MapMergeTypes<Object, Object>, Object> mergePolicy,
                          CallerProvenance provenance) {
@@ -1108,7 +1109,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
                     now, null, persist, true, false);
         }
 
-        return newValue != null;
+        return true;
     }
 
     @Override

